### PR TITLE
Integrate WhiteSource

### PR DESCRIPTION
**Pull request by zkdev**
**What this PR does / why we need it**:
Based on provided component descriptor components are pulled, packed to archives and posted to an whitesource-api-extension endpoint to execute code scans and upload results the [WSS](https://saas.whitesourcesoftware.com/Wss/WSS.html#!home).

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```improvement operator
NONE
```

### CODE REVIEW

File: concourse/steps/scan_sources.py [hunk 1]
```diff
@@ -183,3 +193,45 @@ def scan_sources_and_notify(
...
+            commit_hash = product.util.guess_commit_from_ref(component)
+
+            # download whitesource component
+            whitesource.component.download_component(github_api=wsc.github_api,
+                                                     component_name=wsc.component_name,
+                                                     dest=tmp_file,
+                                                     ref=commit_hash)
+
+            # POST component>
+            res = wsc.whitesource_client.post_component(product_token=wsc.product_token,
```
Comment (zkdev) [inline concourse/steps/scan_sources.py:232]: `POST product`

File: concourse/steps/scan_sources.py [hunk 2]
```diff
@@ -183,3 +193,45 @@ def scan_sources_and_notify(
...
+                                                        component_name=wsc.component_name.name(),
+                                                        requester_email=requester_mail,
+                                                        optional_config=optional,
+                                                        component=tmp_file.name)
```
Comment (zkdev) [inline concourse/steps/scan_sources.py:236]: `archive_path`

File: model/whitesource.py [hunk 3]
```diff
@@ -31,8 +31,14 @@ class WhitesourceConfig(NamedModelElement):
     def base_url(self):
         return self.raw.get('base_url')
```
Comment (zkdev) [inline model/whitesource.py:32]: introduce second config object for extension

File: whitesource/client.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import json
```
Comment (zkdev) [inline whitesource/client.py:1]: pep8 sorting

File: whitesource/component.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    return WhitesourceComponent(
+        whitesource_client=whitesource_client,
+        github_api=github_api,
+        product_token=product_token,
+        component=component,
+        component_name=component_name
+    )
+
+
+class WhitesourceComponent:
```
Comment (zkdev) [inline whitesource/component.py:29]: dataclass
e.g. `UploadRequestObject`

File: concourse/steps/scan_sources.py [hunk 6]
```diff
@@ -183,3 +193,45 @@ def scan_sources_and_notify(
...
+    # for component in []: "components matching component_descriptor"
+    for component in component_descriptor.components():
+
+        # create whitesource component
+        wsc = create_whitesource_component(whitesource_client=client,
+                                           product_token=product_token,
+                                           component=component)
+
+        # store in tmp file
+        with tempfile.NamedTemporaryFile(suffix='.tar.gz') as tmp_file:
```
Comment (zkdev) [inline concourse/steps/scan_sources.py:220]: use anonym tmpfile and pass object directly

**Comment by gardener-robot-ci-3** [comment]
Thank you @zkdev for updating the PR. I will start a build. Once started, the build URL will be posted here.

**Comment by gardener-robot-ci-2** [comment]
Thank you @zkdev for updating the PR. I will start a build. Once started, the build URL will be posted here.

**Comment by gardener-robot-ci-3** [comment]
A build of this pull request has started. You can check on its progress here: https://concourse.ci.gardener.cloud/teams/gardener/pipelines/cc-utils-master/jobs/master-pull-request-job/builds/14

**Pull request merged**